### PR TITLE
Smoothed Bhargava metabolics model

### PR DESCRIPTION
**Pull request by nickbianco**
### Brief summary of changes

This PR brings over the smoothed Bhargava metabolics model and its accompanying C++ example developed by @antoinefalisse from the Moco repository. A summary of the review process for these changes change be found here: opensim-org/opensim-moco#532.

### Testing I've completed

Added test `testMocoMetabolics.cpp`.

### Looking for feedback on...

I think the main point for discussion here is the naming of the component. Currently it is `Bhargava2004Metabolics`, but that might not effectively distinguish itself from `Bhargava2004MuscleMetabolicsProbe`. Perhaps `SmoothBhargava2004Metabolics` or `SmoothBhargava2004MuscleMetabolics` (since it is not really a `Probe`), could be better.

### CHANGELOG.md (choose one)

- updated, but not sure how we're formatting for the next OpenSim release (i.e., 4.2.1 vs 4.3).

---
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2940)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,9 @@ request related to the change, then we may provide the commit.
 
 This is not a comprehensive list of changes but rather a hand-curated collection of the more notable ones. For a comprehensive history, see the [OpenSim Core GitHub repo](https://github.com/opensim-org/opensim-core).
 
+v4.2.1
+======
+- Added Bhargava2004Metabolics, a smoothed version of the Bhargava metabolics model designed for gradient-based optimization (i.e., Moco).
```
Comment (aymanhab) [inline CHANGELOG.md:11]: Please move to 4.2 as not yet finalized

Comment (nickbianco) [reply CHANGELOG.md]: Done.

**Review by aymanhab — CHANGES_REQUESTED** [review]
> Reviewed 8 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2940)* status: 7 of 15 files reviewed, 2 unresolved discussions (waiting on @aymanhab and @nickbianco)

___
*[Bindings/OpenSimHeaders_simulation.h, line 44 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-MTlBEoFE3rOgL-qpMFY:-MTlBEoGCgeQSpOhINow:bdywe9m) ([raw file](https://github.com/opensim-org/opensim-core/blob/96d06749bf36e1b9f4158c271482e98474461d68/Bindings/OpenSimHeaders_simulation.h#L44)):*
> ```objc
> #include <OpenSim/Simulation/Model/Bhargava2004MuscleMetabolicsProbe.h>
> #include <OpenSim/Simulation/Model/Bhargava2004Metabolics.h>
> #include <OpenSim/Simulation/Model/Umberger2010MuscleMetabolicsProbe.h>
> ```

I worry that the class name does not indicate what this object is, other classes in file indicate a Probe, Actuator etc. which are helpful in figuring out where to place in XML files. Is it too late to change?  How do users choose and why?

**Review by nickbianco — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2940)* status: 7 of 15 files reviewed, 2 unresolved discussions (waiting on @aymanhab)

___
*[Bindings/OpenSimHeaders_simulation.h, line 44 at r1](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-MTlBEoFE3rOgL-qpMFY:-MTrFJ1j8na7j8nA6dbQ:b-cvbkdj) ([raw file](https://github.com/opensim-org/opensim-core/blob/96d06749bf36e1b9f4158c271482e98474461d68/Bindings/OpenSimHeaders_simulation.h#L44)):*
<details><summary><i>Previously, aymanhab (Ayman Habib) wrote…</i></summary><blockquote>

I worry that the class name does not indicate what this object is, other classes in file indicate a Probe, Actuator etc. which are helpful in figuring out where to place in XML files. Is it too late to change?  How do users choose and why?
</blockquote></details>

What do you think about `SmoothBhargava2004MuscleMetabolics`, or `SmoothBhargava2004MuscleMetabolicsModel`? It's just a component that contains the metabolics model and provides outputs to compute total metabolics for all muscles associated with the component.

**Comment by tkuchida** [comment]
> What do you think about SmoothBhargava2004MuscleMetabolics, or SmoothBhargava2004MuscleMetabolicsModel? It's just a component that contains the metabolics model and provides outputs to compute total metabolics for all muscles associated with the component.

Perhaps a name that begins with `Bhargava2004` to emphasize its relationship to the existing `Bhargava2004MuscleMetabolicsProbe` class (and to the source paper)? `Bhargava2004SmoothedMuscleMetabolics`?

**Comment by aymanhab** [comment]
My inclination would be to include Bhargava2004 and Smoothed in the name (for the distinction) but also some description of what this is e.g Component or Calculator or MocoGoalTerm etc. so you end up with something like SmoothedBhargava2004MetabolicsCalculator or SmoothedBhargava2004MetabolicsComponent or ....? Thoughts?

**Comment by nickbianco** [comment]
I like @tkuchida's suggestion: `Bhargava2004SmoothedMuscleMetabolics`. Not sure that we need to end with `Calculator` or `Component`, both seem a little redundant.

**Comment by chrisdembia** [comment]
How about Bhargava2004SmoothedMuscleMetabolicsForGradientBasedOptimization?

**Comment by aymanhab** [comment]
> How about Bhargava2004SmoothedMuscleMetabolicsForGradientBasedOptimization?

Sounds great if not too long. The issue I had originally that was not addressed by earlier proposals is how to use it, if it is a cost term, we can append Cost instead of 'ForGradientBasedOptimization' if that's consistent with Moco's terminology

**Comment by chrisdembia** [comment]
Oh sorry that was a joke

**Comment by aymanhab** [comment]
@carmichaelong @antoinefalisse @jenhicks are the current names good enough for you to address the concerns raised earlier in the review? If so I'm ok with renaming and moving on if not I'd like to hear your thoughts. Thank you.

**Comment by antoinefalisse** [comment]
I like `Bhargava2004SmoothedMuscleMetabolics` the most (it is not a `MocoGoal` by itself). Thanks!

**Comment by carmichaelong** [comment]
+1 for `Bhargava2004SmoothedMuscleMetabolics`. I especially like that it starts with `Bhargava2004` (which follows other naming schemes like the previous probes mentioned and `Millard2012...` `Thelen2003...`) as @tkuchida emphasized

**Comment by aymanhab** [comment]
Thanks all for the feedback, can you apply the change @nickbianco so we move on? Thank you

**Comment by nickbianco** [comment]
@aymanhab Ready for review.

**Review by aymanhab — CHANGES_REQUESTED** [review]
> Reviewed 1 of 15 files at r1, 9 of 9 files at r3.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2940)* status: all files reviewed, 4 unresolved discussions (waiting on @aymanhab and @nickbianco)

___
*[OpenSim/Moco/Test/testMocoMetabolics.cpp, line 81 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-MUAxcds4-K0UOKjDsZI:-MUAxcds4-K0UOKjDsZJ:bjuouxb) ([raw file](https://github.com/opensim-org/opensim-core/blob/292bbec69bd155c1712bfc4800e32eed6997c3c5/OpenSim/Moco/Test/testMocoMetabolics.cpp#L81)):*
> ```cpp
>     metabolicsPtr_smooth_tanh->setName("metabolics_smooth_tanh");
>     metabolicsPtr_smooth_tanh->set_use_smoothing(true);
>     // We set a high value for the velocity and heat rate smoothing parameters
> ```

I'd expect "Smoothed" class to have "use_smoothing" set to true be default, and that you turn it off to get non-smooth behavior, does that make sense?
___
*[OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h, line 171 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-MUAyx1TD7jMKzeP9Pft:-MUAyx1TD7jMKzeP9Pfu:b-f0c3p2) ([raw file](https://github.com/opensim-org/opensim-core/blob/292bbec69bd155c1712bfc4800e32eed6997c3c5/OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h#L171)):*
> ```objc
> /// discontinuous function have successfully converged; therefore, we have
> /// included it in this implementation of the model.
> ///
> ```

Is there a description somewhere to help new users figure out how to use this object, in addition to all the details about the internals of the implementation?
Also why is this a ModelComponent as opposed to a generic Component?
___
*[OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h, line 251 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-MUAzqIqE-NZGzOAqrE_:-MUAzqIr1crL0l13UHE8:b-mlenni) ([raw file](https://github.com/opensim-org/opensim-core/blob/292bbec69bd155c1712bfc4800e32eed6997c3c5/OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h#L251)):*
> ```objc
> 
>     int getNumMetabolicMuscles() const;
> ```

Some comment about default values assumed when not passed in to constructors.

**Review by aymanhab — COMMENTED** [review]
Thanks @nickbianco LGTM module some more documentation.  Do you expect this class to be available through scripting? I don't see it included in swig .i files

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-:-MUB-Fxl3gHHLeBhYFKv:bg9netz)* status: all files reviewed, 4 unresolved discussions (waiting on @aymanhab and @nickbianco)

**Review by nickbianco — COMMENTED** [review]
The model is available in scripting, see changes to `simulation.i` and `OpenSimHeaders_simultion.h`.

> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-:-MUBR2v1EslUXBCMNTsh:b97uxk2)* status: 14 of 15 files reviewed, 3 unresolved discussions (waiting on @aymanhab)

___
*[OpenSim/Moco/Test/testMocoMetabolics.cpp, line 81 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-MUAxcds4-K0UOKjDsZI:-MUBRDZH6JCZ-AF6ZskE:b5bguus) ([raw file](https://github.com/opensim-org/opensim-core/blob/292bbec69bd155c1712bfc4800e32eed6997c3c5/OpenSim/Moco/Test/testMocoMetabolics.cpp#L81)):*
<details><summary><i>Previously, aymanhab (Ayman Habib) wrote…</i></summary><blockquote>

I'd expect "Smoothed" class to have "use_smoothing" set to true be default, and that you turn it off to get non-smooth behavior, does that make sense?
</blockquote></details>

As the non-smoothed option is technically the more faithful version of the model,  I think we opted to turn off the smoothing as a default. Also, turning on the smoothing via the property forces the user to "opt-in" to using smoothing, which is an intended feature.

I added a comment to the component docs to point users to the `use_smoothing` property.
___
*[OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h, line 171 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-MUAyx1TD7jMKzeP9Pft:-MUBJKsJ8ctzTXNtF1hU:baaewrh) ([raw file](https://github.com/opensim-org/opensim-core/blob/292bbec69bd155c1712bfc4800e32eed6997c3c5/OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h#L171)):*
<details><summary><i>Previously, aymanhab (Ayman Habib) wrote…</i></summary><blockquote>

Is there a description somewhere to help new users figure out how to use this object, in addition to all the details about the internals of the implementation?
Also why is this a ModelComponent as opposed to a generic Component?
</blockquote></details>

`example2DWalkingMetabolics.cpp`, which was added in the PR, is the best reference for how to use the model, but I've added some more description here as well. 

I believe we needed to use ModelComponent such that the Sockets to the muscles added via `addMuscle()` were not lost when constructing the model.
___
*[OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h, line 251 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-MUAzqIqE-NZGzOAqrE_:-MUBQwCq4gjUAZ4_WtRe:bp3gw76) ([raw file](https://github.com/opensim-org/opensim-core/blob/292bbec69bd155c1712bfc4800e32eed6997c3c5/OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h#L251)):*
<details><summary><i>Previously, aymanhab (Ayman Habib) wrote…</i></summary><blockquote>

Some comment about default values assumed when not passed in to constructors.
</blockquote></details>

I added a note about default values in the component docs above.

**Comment by nickbianco** [comment]
@aymanhab Ready for review.

**Review by aymanhab — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2940)* status: all files reviewed, 1 unresolved discussion (waiting on @nickbianco)

___
*[OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h, line 171 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-MUAyx1TD7jMKzeP9Pft:-MUExRNx1YrH0JdvPPaD:bwvn00e) ([raw file](https://github.com/opensim-org/opensim-core/blob/292bbec69bd155c1712bfc4800e32eed6997c3c5/OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h#L171)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`example2DWalkingMetabolics.cpp`, which was added in the PR, is the best reference for how to use the model, but I've added some more description here as well. 

I believe we needed to use ModelComponent such that the Sockets to the muscles added via `addMuscle()` were not lost when constructing the model.


</blockquote></details>

Thanks @nickbianco Examples are great but will not help somebody looking cold at the documentation/doxygen, that's why a code snippet or XML segment is preferred. Maybe a link to the cpp file would do.

**Review by nickbianco — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2940)* status: all files reviewed, 1 unresolved discussion (waiting on @aymanhab)

___
*[OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h, line 171 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/2940#-MUAyx1TD7jMKzeP9Pft:-MUFLSemCUuhLXFjCAks:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/292bbec69bd155c1712bfc4800e32eed6997c3c5/OpenSim/Simulation/Model/Bhargava2004SmoothedMuscleMetabolics.h#L171)):*
<details><summary><i>Previously, aymanhab (Ayman Habib) wrote…</i></summary><blockquote>

Thanks @nickbianco Examples are great but will not help somebody looking cold at the documentation/doxygen, that's why a code snippet or XML segment is preferred. Maybe a link to the cpp file would do.
</blockquote></details>

Done.

**Review by aymanhab — APPROVED** [review]
Excellent, thanks @nickbianco LGTM 👍

**Pull request merged**